doc/editorsettings/editorsettings.md
IDE Customization
=================

Codimension supports a number of customizable parameters via editing
the skin files. See the documentation [here](../colorschemes/colorschemes.md).

On top of it Codimension has a UI for changing some of the most frequently
used settings, including the editor settings. The UI is available in the
IDE main menu -> Options

![IDE Options](options.png "IDE Options")

All the slectable options are radio or check box choices. Selecting any of them
(except of a skin) affects the IDE immediately.

| Option | Description |
|:-------|:------------|
| Show vertical edge | To show the vertical edge in the text editor at certain character. Default: 80. Can be changed via editing `~/.codimension3/settings.json` |
| Show whitespaces | To show the whitespace characters in the text editor unless it is exactly one character between words |
| Wrap long lines | If a line does not fit the available width then it will be wrapped to the next one (display only) |
| Show brace matching | Highlight braces when the text cursor is on one of them |
| Show indentation guides | To show the indentation guides (only if whitespace characters are suppressed) |
| Highlight current line | To highlight the line where the cursor is |
| HOME to first non space | To move the text cursor to the first non space character when the HOME key is pressed |
| Auto remove trailing spaces on save | Automatically remove the trailing spaces when the file is saved |
| Editor calltips | To show the editor calltips |
| Show navigation bar | To show the text editor navigation bar |
| Show control flow navigation bar | To show the control flow navigation bar |
| Show main toolbar | To show the main toolbar |
| Redirected I/O | The submenu to select the behavior of the redirected I/O concole tab at the bottom |
| Tooltips | The submenu to select what tooltips the IDE will display |
| Open tabs button | The submenu to specify the IDE behavior when the 'Open tabs' button is clicked and an item is selected |
| Skins | The submenu to select the IDE skin (requires IDE restart) |
| Styles | The submenu to select the QT library widgets appearence |
| Text mono font family | The submenu to select the monospace font face |
| Editor margin font family | The submenu to select the editor margin font face |
| Flow mono font family | The submenu to select the graphics pane font face |
| Flow badge font family | The submenu to select the graphics pane scope item badge font face |

The submenues to select fonts are populated dynamically depending on what fonts
are available on a particular system. On a Linux system users may add fonts so
that they are available for them in Codimension by copying the font files into
their `~/.fonts` directory. Here are a few links to fonts which may be of
interest:
- [Iosevka](https://github.com/be5invis/Iosevka) monospace font
- [Ubuntu](http://font.ubuntu.com/) mono font
- [Cascadia](https://github.com/microsoft/cascadia-code) monospace font



ChangeLog
2019 xxx xx Sergey Satskiy <sergey.satskiy@tuta.io>

* Version x.x.x
* Bug fix: allow grouping for the try-except blocks
* Bug fix: exception when selecting doc links
* CML doc support
* Embedded documentation significant update

2019 Aug 05 Sergey Satskiy <sergey.satskiy@tuta.io>

* Version 4.5.2
* Bug fix: Exception when switching smart zoom level

2019 Jul 31 Sergey Satskiy <sergey.satskiy@tuta.io>

* Version 4.5.1
* PlantUML diagram support for markdown
* Self protection against invalid custom terminal command line
* Bumping up the cdmcfparser version to 2.2.1
* Bumping up the cdmpyparser version to 3.1.1
* Bumping up the jedi library version to 0.14.1

2019 Apr 25 Sergey Satskiy <sergey.satskiy@gmail.com>

* Version 4.5.0
* Python 3.7 support
* New feature: support a floating independent window with flowchart renderer
* More on embedded documentation
* Bumping up the cdmcfparser version to 2.2.0
* Bumping up the cdmpyparser version to 3.1.0

2019 Mar 21 Sergey Satskiy <sergey.satskiy@gmail.com>

* Version 4.4.0
* Bumping up the magic library version to 0.4.15
* Introducing basic markdown rendering support
* Bug fix: proper file position is saved when an individual tab is closed
* Improved file type detection for some file types
* Bug fix: incorrect inital margin size when a new tab button is clicked
* Bumping up the jedi library version to 0.13.1
* Bumping up the radon library version to 2.4.0
* Fixing the flake8 version to 3.6.0
* Support embedded IDE documentation in Markdown
* Bug fix: running project profile instead of debug
* Bug fix: incorrect project file encoding detection if a file did not have an
  explicit encoding
* Bug fix: gprof2dot interface has been updated in a non-backward compatible
  manner

2018 Oct 02 Sergey Satskiy <sergey.satskiy@gmail.com>

* Version 4.3.1
* Bug fix: crash when a try else block is in an if branch and a mouse hovers it
* Downgrading pyflakes requirement to 1.6.0 due to an implicit dependency via
  radon->flake8_polyfill->flake8->pyflakes

2018 Sep 26 Sergey Satskiy <sergey.satskiy@gmail.com>

* Version 4.3.0
* Bump jedi version to 0.12.1
* Buffer cyclomatic complexity on-the-fly: status bar menu and margin marks
* Better analysis margin pixmap scaling
* Print button for the text editor toolbar added
* Integrate vulture to find dead code
* Remove obsolete unreliable unused functions/classes/global variable
  search implementation based on jedi
* Bug fix: crash on debug dump settings from the main menu
* New dependencies (radon - complexity, vulture - dead code analysis)
* Minor improvements

2018 Aug 13 Sergey Satskiy <sergey.satskiy@gmail.com>

* Version 4.2.4
* Fix: default application CSS adjustments
* Rubber band selection on the graphics pane

2018 May 25 Sergey Satskiy <sergey.satskiy@gmail.com>

* Version 4.2.3
* Bug fix: IDE does not start after installation (second try)

2018 May 17 Sergey Satskiy <sergey.satskiy@gmail.com>

* Version 4.2.2
* Bug fix: IDE does not start after installation
* Preliminary rubber band UI on the graphics pane

2018 May 11 Sergey Satskiy <sergey.satskiy@gmail.com>

* Version 4.2.1
* Bug fix: incorrect recent files loading (regression)
* New feature: grouping items on the graphics pane

2018 Apr 29 Sergey Satskiy <sergey.satskiy@gmail.com>

* Version 4.2.0
* Text-to-graphics group support: empty/collapsed/expanded groups.
* Bug fix: cloning of a tab content
* Bug fix: incorrect saving encoding when a new file overwrites an existing one
* Bug fix: incorrect item width on smart zoom 1 when scale is changed
* Recent files list is populated when there is no project
* Minor improvements

2018 Jan 11 Sergey Satskiy <sergey.satskiy@gmail.com>

* Version 4.1.0
* Proportional change of spacings when a mono font changed for the flow UI
* New feature: show/hide comments/docstrings/except blocks on the flowUI
* More flexible pyflakes suppression line
* New feature: smart zoom; four levels are available now:
  - full view
  - no content, no comment, no docstring
  - no content, no comment, no docstring, no plain items, no decorators
  - only top level classes and functions
* Minor bug fixes

2017 Dec 3 Sergey Satskiy <sergey.satskiy@gmail.com>

* Verion 4.0.4
* Better approach to setup.py (on some systems it did not work as expected)

2017 Nov 30 Sergey Satskiy <sergey.satskiy@gmail.com>

* Version 4.0.3
* Cosmetic changes (pep8 compatibility)
* New flow setting to enable CML indicators on the diagram (default: False)
* Better icon for the 'build imports diagram' buttons/menu items
* Context menu on the flow UI to switch if branches when many of them are
  selected
* Multiple bug fixes
* UI for the control flow to replace text and for custom colors

2017 Nov 19 Sergey Satskiy <sergey.satskiy@gmail.com>

* Version 4.0.2
* reworked import resolution procedure
* bug fix: imports/definitions panel resizing when a text editor is resized
* bug fix: do not use native dialogs to open/save files

2017 Nov 02 Sergey Satskiy <sergey.satskiy@gmail.com>

* Version 4.0.1
* Removing debugging printouts
* pypi friendlier package description

2017 Oct 29 Sergey Satskiy <sergey.satskiy@gmail.com>

* Version 4.0.0
* First public version for Python 3. From now on Python 2 is not supported.
* pip installable

2016 Dec 26 Sergey Satskiy <sergey.satskiy@gmail.com>

* Version 3.0.2
* Getting rid of QWebKit dependency. It is not found in newer Ubuntu releases


codimension/utils/settings.py
# -*- coding: utf-8 -*-
#
# codimension - graphics python two-way code editor and analyzer
# Copyright (C) 2010-2016  Sergey Satskiy <sergey.satskiy@gmail.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

"""codimension settings"""

import os
import os.path
import datetime
import json
import logging
from copy import deepcopy
from ui.qt import QObject, QDir, pyqtSignal
from .config import SETTINGS_ENCODING, CONFIG_DIR
from .runparamscache import RunParametersCache
from .debugenv import DebuggerEnvironment
from .searchenv import SearchEnvironment
from .fsenv import FileSystemEnvironment
from .filepositions import FilePositions
from .userencodings import FileEncodings
from .flowgroups import FlowUICollapsedGroups
from .webresourcecache import WebResourceCache
from .plantumlcache import PlantUMLCache


SETTINGS_DIR = os.path.join(os.path.realpath(QDir.homePath()),
                            CONFIG_DIR) + os.path.sep

CLEAR_AND_REUSE = 0
NO_CLEAR_AND_REUSE = 1
NO_REUSE = 2


class ProfilerSettings:

    """Holds IDE-wide profiler options"""

    def __init__(self):
        self.nodeLimit = 1.0
        self.edgeLimit = 1.0

    def profSettingsToJSON(self):
        """Converts the instance to a serializable structure"""
        return {'__class__': 'ProfilerSettings',
                '__values__': {'nodeLimit': self.nodeLimit,
                               'edgeLimit': self.edgeLimit}}

    def profSettingsFromJSON(self, jsonObj):
        """Populates the values from the json object"""
        self.nodeLimit = jsonObj['__values__']['nodeLimit']
        self.edgeLimit = jsonObj['__values__']['edgeLimit']

    def __eq__(self, other):
        return self.nodeLimit == other.nodeLimit and \
               self.edgeLimit == other.edgeLimit


class DebuggerSettings:

    """Holds IDE-wide debugger options"""

    def __init__(self):
        self.reportExceptions = True
        self.traceInterpreter = True
        self.stopAtFirstLine = True
        self.autofork = False
        self.followChild = False

    def debugSettingsToJSON(self):
        """Converts the instance to a serializable structure"""
        return {'__class__': 'DebuggerSettings',
                '__values__': {'reportExceptions': self.reportExceptions,
                               'traceInterpreter': self.traceInterpreter,
                               'stopAtFirstLine': self.stopAtFirstLine,
                               'autofork': self.autofork,
                               'followChild': self.followChild}}

    def debugSettingsFromJSON(self, jsonObj):
        """Populates the values from the json object"""
        self.reportExceptions = jsonObj['__values__']['reportExceptions']
        self.traceInterpreter = jsonObj['__values__']['traceInterpreter']
        self.stopAtFirstLine = jsonObj['__values__']['stopAtFirstLine']
        self.autofork = jsonObj['__values__']['autofork']
        self.followChild = jsonObj['__values__']['followChild']

    def __eq__(self, other):
        return self.reportExceptions == other.reportExceptions and \
               self.traceInterpreter == other.traceInterpreter and \
               self.stopAtFirstLine == other.stopAtFirstLine and \
               self.autofork == other.autofork and \
               self.followChild == other.followChild


_DEFAULT_SETTINGS = {
    # general
    'zoom': 0,
    'flowZoom': 0,
    'smartZoom': 0,
    'xpos': 50,
    'ypos': 50,
    'width': 750,
    'height': 550,
    'rendererxpos': 425,
    'rendererypos': 75,
    'rendererwidth': 375,
    'rendererheight': 550,
    'screenwidth': 0,
    'screenheight': 0,
    'xdelta': 0,
    'ydelta': 0,
    'skin': 'default',
    'modifiedFormat': '%s *',
    'verticalEdge': True,
    'showSpaces': True,
    'lineWrap': False,
    'showBraceMatch': True,
    'indentationGuides': False,
    'currentLineVisible': True,
    'jumpToFirstNonSpace': False,
    'removeTrailingOnSave': False,
    'showFSViewer': True,
    'showStackViewer': True,
    'showThreadViewer': True,
    'showIgnoredExcViewer': True,
    'showWatchPointViewer': True,
    'showNavigationBar': True,
    'showCFNavigationBar': True,
    'showMainToolBar': True,
    'profilerLimits': ProfilerSettings(),
    'debuggerSettings': DebuggerSettings(),
    'debugGLFilter': 0,
    'editorEdge': 80,
    'projectTooltips': True,
    'recentTooltips': True,
    'classesTooltips': True,
    'functionsTooltips': True,
    'outlineTooltips': True,
    'findNameTooltips': True,
    'findFileTooltips': True,
    'editorTooltips': True,
    'editorCalltips': True,
    'leftBarMinimized': False,
    'bottomBarMinimized': False,
    'rightBarMinimized': False,
    'projectLoaded': False,
    'clearDebugIO': False,
    'floatingRenderer': False,
    'hSplitterSizes': [200, 450, 575],
    'vSplitterSizes': [400, 150],
    'flowSplitterSizes': [225, 225],
    'style': 'fusion',
    'vcsstatusupdateinterval': 30,      # seconds
    'tablistsortalpha': True,
    'taborderpreserved': False,
    'maxRecentProjects': 32,
    'maxRecentFiles': 32,
    'maxSearchEntries': 32,
    'maxHighlightedMatches': 256,
    'maxBreakpoints': 63,               # per file
    'encoding': 'utf-8',
    'hidedocstrings': False,
    'hidecomments': False,
    'hideexcepts': False,

    # Debug variable filters
    'dbgfltlocal': True,
    'dbgfltglobal': True,
    'dbgflthidden': True,
    'dbgflttype': False,
    'dbgfltmethod': False,
    'dbgfltfunc': False,
    'dbgfltbuiltin': False,
    'dbgfltmodule': False,
    'dbgfltnotype': False,

    'calltrace': True,

    # The IO redirect console
    'ioconsolemaxmsgs': 10000,
    'ioconsoledelchunk': 512,
    'ioconsolelinewrap': False,
    'ioconsoleshowspaces': True,
    'ioconsoleautoscroll': True,
    'ioconsolereuse': CLEAR_AND_REUSE,

    'navbarglobalsimports': False,

    'recentProjects': [],
    'projectFilesFilters': ['^\\.', '.*\\~$', '.*\\.pyc$',
                            '.*\\.swp$', '.*\\.pyo$', '__pycache__'],
    'ignoredExceptions': [],
    'disabledPlugins': [],
    'vcsindicators': [[-1, 'vcsunversioned.png', None,
                       '220,220,255,255', 'Not under VCS control'],
                      [-2, 'vcsstatuserror.png', None,
                       '255,160,160,255', 'Error getting status']]}


def settingsFromJSON(jsonObj):
    """Custom deserialization"""
    if '__class__' in jsonObj:
        if jsonObj['__class__'] == 'ProfilerSettings':
            pSettings = ProfilerSettings()
            pSettings.profSettingsFromJSON(jsonObj)
            return pSettings
        if jsonObj['__class__'] == 'DebuggerSettings':
            dSettings = DebuggerSettings()
            dSettings.debugSettingsFromJSON(jsonObj)
            return dSettings
    return jsonObj


def settingsToJSON(pythonObj):
    """Custom serialization"""
    if isinstance(pythonObj, ProfilerSettings):
        return pythonObj.profSettingsToJSON()
    if isinstance(pythonObj, DebuggerSettings):
        return pythonObj.debugSettingsToJSON()
    raise TypeError(repr(pythonObj) + ' is not JSON serializable')


class SettingsWrapper(QObject,
                      DebuggerEnvironment,
                      SearchEnvironment,
                      FileSystemEnvironment,
                      RunParametersCache,
                      FilePositions,
                      FileEncodings,
                      FlowUICollapsedGroups):

    """Provides settings singleton facility"""

    MAX_SMART_ZOOM = 3

    sigRecentListChanged = pyqtSignal()
    sigFlowSplitterChanged = pyqtSignal()
    sigFlowZoomChanged = pyqtSignal()
    sigTextZoomChanged = pyqtSignal()
    sigHideDocstringsChanged = pyqtSignal()
    sigHideCommentsChanged = pyqtSignal()
    sigHideExceptsChanged = pyqtSignal()
    sigSmartZoomChanged = pyqtSignal()
    sigRecentFilesChanged = pyqtSignal()

    def __init__(self):
        QObject.__init__(self)
        DebuggerEnvironment.__init__(self)
        SearchEnvironment.__init__(self)
        FileSystemEnvironment.__init__(self)
        RunParametersCache.__init__(self)
        FilePositions.__init__(self)
        FileEncodings.__init__(self)
        FlowUICollapsedGroups.__init__(self)

        self.minTextZoom = None
        self.minCFlowZoom = None

        self.__values = deepcopy(_DEFAULT_SETTINGS)

        # make sure that the directory exists
        if not os.path.exists(SETTINGS_DIR):
            os.makedirs(SETTINGS_DIR)

        RunParametersCache.setup(self, SETTINGS_DIR)
        DebuggerEnvironment.setup(self, SETTINGS_DIR)
        SearchEnvironment.setup(self, SETTINGS_DIR)
        FileSystemEnvironment.setup(self, SETTINGS_DIR)
        FilePositions.setup(self, SETTINGS_DIR)
        FileEncodings.setup(self, SETTINGS_DIR)
        FlowUICollapsedGroups.setup(self, SETTINGS_DIR)

        self.webResourceCache = WebResourceCache(SETTINGS_DIR + os.path.sep +
                                                 'webresourcecache')
        self.plantUMLCache = PlantUMLCache(SETTINGS_DIR + os.path.sep +
                                           'plantumlcache')

        # Save the config file name
        self.__fullFileName = SETTINGS_DIR + "settings.json"

        # Create file if does not exist
        if not os.path.exists(self.__fullFileName):
            # Save to file
            self.flush()
            return

        readErrors = []

        # Load the previous session settings
        try:
            with open(self.__fullFileName, "r",
                      encoding=SETTINGS_ENCODING) as diskfile:
                diskValues = json.load(diskfile, object_hook=settingsFromJSON)
        except Exception as exc:
            # Bad error - save default
            self.__saveErrors('Could not read setting from ' +
                              self.__fullFileName + ': ' + str(exc) +
                              'Overwriting with the default settings...')
            self.flush()
            return

        for item, val in diskValues.items():
            if item in self.__values:
                if type(self.__values[item]) != type(val):
                    readErrors.append("Settings '" + item +
                                      "' type from the disk file " +
                                      self.__fullFileName +
                                      ' does not match the expected one. '
                                      'The default value is used.')
                else:
                    self.__values[item] = val
            else:
                readErrors.append('Disk file ' + self.__fullFileName +
                                  " contains extra value '" + item +
                                  "'. It will be lost.")

        # If format is bad then overwrite the file
        if readErrors:
            self.__saveErrors("\n".join(readErrors))
            self.flush()

        SearchEnvironment.setLimit(self, self.__values['maxSearchEntries'])
        FileSystemEnvironment.setLimit(self, self.__values['maxRecentFiles'])

    @staticmethod
    def __saveErrors(message):
        """Appends the message to the startup errors file"""
        try:
            with open(SETTINGS_DIR + 'startupmessages.log', 'a',
                      encoding=SETTINGS_ENCODING) as diskfile:
                diskfile.write('------ Startup report at ' +
                               str(datetime.datetime.now()) + '\n')
                diskfile.write(message)
                diskfile.write('\n------\n\n')
        except:
            # This is not that important
            pass

    def flush(self):
        """Writes the settings to the disk"""
        try:
            with open(self.__fullFileName, 'w',
                      encoding=SETTINGS_ENCODING) as diskfile:
                json.dump(self.__values, diskfile,
                          default=settingsToJSON, indent=4)
        except Exception as exc:
            logging.error('Errol saving setting (to ' + self.__fullFileName +
                          '): ' + str(exc))

    def addRecentProject(self, projectFile, needFlush=True):
        """Adds the recent project to the list"""
        absProjectFile = os.path.realpath(projectFile)
        recentProjects = self.__values['recentProjects']

        if absProjectFile in recentProjects:
            recentProjects.remove(absProjectFile)

        recentProjects.insert(0, absProjectFile)
        limit = self.__values['maxRecentProjects']
        if len(recentProjects) > limit:
            recentProjects = recentProjects[0:limit]
        self.__values['recentProjects'] = recentProjects
        if needFlush:
            self.flush()
        self.sigRecentListChanged.emit()

    def deleteRecentProject(self, projectFile, needFlush=True):
        """Deletes the recent project from the list"""
        absProjectFile = os.path.realpath(projectFile)
        recentProjects = self.__values['recentProjects']

        if absProjectFile in recentProjects:
            recentProjects.remove(absProjectFile)
            self.__values['recentProjects'] = recentProjects
            if needFlush:
                self.flush()
            self.sigRecentListChanged.emit()

    @staticmethod
    def getDefaultGeometry():
        """Provides the default window size and location"""
        return _DEFAULT_SETTINGS['xpos'], _DEFAULT_SETTINGS['ypos'], \
               _DEFAULT_SETTINGS['width'], _DEFAULT_SETTINGS['height']

    @staticmethod
    def getDefaultRendererWindowGeometry():
        """Provides the default renderer window size and location"""
        # A bit shifted down and half of the width of the main window
        return _DEFAULT_SETTINGS['rendererxpos'], \
               _DEFAULT_SETTINGS['rendererypos'], \
               _DEFAULT_SETTINGS['rendererwidth'], \
               _DEFAULT_SETTINGS['rendererheight']

    def getProfilerSettings(self):
        """Provides the profiler IDE-wide settings"""
        return self.__values['profilerLimits']

    def setProfilerSettings(self, newValue, needFlush=True):
        """Updates the profiler settings"""
        if self.__values['profilerLimits'] != newValue:
            self.__values['profilerLimits'] = newValue
            if needFlush:
                self.flush()

    def getDebuggerSettings(self):
        """Provides the debugger IDE-wide settings"""
        return self.__values['debuggerSettings']

    def setDebuggerSettings(self, newValue, needFlush=True):
        """Updates the debugger settings"""
        if self.__values['debuggerSettings'] != newValue:
            self.__values['debuggerSettings'] = newValue
            if needFlush:
                self.flush()

    def validateZoom(self, minTextZoom, minCFlowZoom):
        """Validates the min zoom values"""
        self.minTextZoom = minTextZoom
        self.minCFlowZoom = minCFlowZoom
        warnings = []
        if self.__values['zoom'] < minTextZoom:
            warnings.append('The current text zoom (' +
                            str(self.__values['zoom']) +
                            ') will be adjusted to ' + str(minTextZoom) +
                            ' due to it is less than min fonts allowed.')
            self.__values['zoom'] = minTextZoom
        if self.__values['flowZoom'] < minCFlowZoom:
            warnings.append('The current flow zoom (' +
                            str(self.__values['flowZoom']) +
                            ') will be adjusted to ' + str(minCFlowZoom) +
                            ' due to it is less than min fonts allowed.')
            self.__values['flowZoom'] = minCFlowZoom
        if self.__values['smartZoom'] < 0:
            warnings.append('The current smart zoom (' +
                            str(self.__values['smartZoom']) +
                            ') will be adjusted to 0 due to it must be >= 0')
            self.__values['smartZoom'] = 0
        elif self.__values['smartZoom'] > SettingsWrapper.MAX_SMART_ZOOM:
            warnings.append('The current smart zoom (' +
                            str(self.__values['smartZoom']) +
                            ') will be adjusted to ' +
                            str(SettingsWrapper.MAX_SMART_ZOOM) +
                            ' due to it is larger than max allowed.')
            self.__values['smartZoom'] = SettingsWrapper.MAX_SMART_ZOOM

        if warnings:
            self.flush()
        return warnings

    def __getitem__(self, key):
        return self.__values[key]

    def __setitem__(self, key, value):
        self.__values[key] = value
        if key == 'flowSplitterSizes':
            self.sigFlowSplitterChanged.emit()
        elif key == 'hidedocstrings':
            self.sigHideDocstringsChanged.emit()
        elif key == 'hidecomments':
            self.sigHideCommentsChanged.emit()
        elif key == 'hideexcepts':
            self.sigHideExceptsChanged.emit()
        self.flush()

    def onTextZoomIn(self):
        """Triggered when the text is zoomed in"""
        self.__values['zoom'] += 1
        self.flush()
        self.sigTextZoomChanged.emit()

    def onTextZoomOut(self):
        """Triggered when the text is zoomed out"""
        if self.__values['zoom'] > self.minTextZoom:
            self.__values['zoom'] -= 1
            self.flush()
            self.sigTextZoomChanged.emit()

    def onTextZoomReset(self):
        """Triggered when the text zoom is reset"""
        if self.__values['zoom'] != 0:
            self.__values['zoom'] = 0
            self.flush()
            self.sigTextZoomChanged.emit()

    def onSmartZoomIn(self):
        """Triggered when the smart zoom is changed"""
        if self.__values['smartZoom'] < SettingsWrapper.MAX_SMART_ZOOM:
            self.__values['smartZoom'] += 1
            self.flush()
            self.sigSmartZoomChanged.emit()

    def onSmartZoomOut(self):
        """Triggered when the smart zoom is changed"""
        if self.__values['smartZoom'] > 0:
            self.__values['smartZoom'] -= 1
            self.flush()
            self.sigSmartZoomChanged.emit()

    def onFlowZoomIn(self):
        """Triggered when the flow is zoomed in"""
        self.__values['flowZoom'] += 1
        self.flush()
        self.sigFlowZoomChanged.emit()

    def onFlowZoomOut(self):
        """Triggered when the flow is zoomed out"""
        if self.__values['flowZoom'] > self.minCFlowZoom:
            self.__values['flowZoom'] -= 1
            self.flush()
            self.sigFlowZoomChanged.emit()

    def onFlowZoomReset(self):
        """Triggered when the flow zoom is reset"""
        if self.__values['flowZoom'] != 0:
            self.__values['flowZoom'] = 0
            self.flush()
            self.sigFlowZoomChanged.emit()

    def addRecentFile(self, path):
        """Adds a recent file. True if a new file was inserted."""
        ret = FileSystemEnvironment.addRecentFile(self, path)
        if ret:
            self.sigRecentFilesChanged.emit()
        return ret


SETTINGS_SINGLETON = SettingsWrapper()


def Settings():
    """Settings singleton access"""
    return SETTINGS_SINGLETON
